mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,350 @@
+    dst_decimals: Option<u8>,
+    /// Chain id where interim LR swap occurs (obtained from the destination token)
+    chain_id: Option<u64>,
+    /// Estimated src token / dst token price
+    lr_price: Option<MmNumber>,
+    /// A quote from LR provider with tx data to do interim LR swap of src token to dst token
+    lr_swap_data: Option<ClassicSwapData>,
+}
+
+impl LrData {
+    #[allow(clippy::result_large_err)]
+    fn get_chain_contract_info(&self) -> MmResult<(String, String, u64), ApiIntegrationRpcError> {
+        let src_contract = self
+            .src_contract
+            .as_ref()
+            .ok_or(ApiIntegrationRpcError::InternalError(
+                "Source LR contract not set".to_owned(),
+            ))?
+            .display_address();
+        let dst_contract = self
+            .dst_contract
+            .as_ref()
+            .ok_or(ApiIntegrationRpcError::InternalError(
+                "Destination LR contract not set".to_owned(),
+            ))?
+            .display_address();
+        let chain_id = self
+            .chain_id
+            .ok_or(ApiIntegrationRpcError::InternalError("LR chain id not set".to_owned()))?;
+        Ok((src_contract, dst_contract, chain_id))
+    }
+}
+
+struct LrDataMap {
+    /// Map to store data needed for best price estimations for swaps with LR,
+    /// the key is the source and destination token pair from the LR swap part
+    inner: HashMap<(Ticker, Ticker), LrData>,
+}
+
+impl LrDataMap {
+    /// Init LR data map from the source token (mytoken) and tokens from orders
+    fn new_with_src_token(src_token: Ticker, orders: Vec<RpcOrderbookEntryV2>) -> Self {
+        Self {
+            inner: orders
+                .into_iter()
+                .map(|order| {
+                    ((src_token.clone(), order.coin.clone()), LrData {
+                        order,
+                        src_contract: None,
+                        src_decimals: None,
+                        src_amount: None,
+                        dst_contract: None,
+                        dst_amount: None,
+                        dst_decimals: None,
+                        chain_id: None,
+                        lr_price: None,
+                        lr_swap_data: None,
+                    })
+                })
+                .collect::<HashMap<(_, _), _>>(),
+        }
+    }
+
+    /// Calculate amounts of destination tokens required to fill ask orders for the requested base_amount:
+    /// multiplies base_amount by the order price (base_amount must be in 'coins')
+    async fn calc_destination_token_amounts(
+        &mut self,
+        ctx: &MmArc,
+        base_amount: &MmNumber,
+    ) -> MmResult<(), ApiIntegrationRpcError> {
+        for lr_data in self.inner.values_mut() {
+            let price: MmNumber = lr_data.order.price.rational.clone().into();
+            let dst_amount = base_amount * &price;
+            let coin = lp_coinfind_or_err(ctx, &lr_data.order.coin).await?;
```

Comment:
also the same: can we gracefully skip this entry?

Reply:
Maybe, I am not sure about this yet.
For now I have followed a principle like 'continue if getting errors from the provider and abort if getting an error in kdf' (considering that our config must be valid).